cmd/labs/project/installer.go

```diff
@@ -0,0 +1,280 @@
+		version:       "latest",
+		gitHubAPI:     "https://api.github.com",
+		gitHubContent: "https://raw.githubusercontent.com",
+	}
+}
+
+type installer struct {
+	cmd           *cobra.Command
+	name          string
+	version       string
+	gitHubAPI     string
+	gitHubContent string
+}
+
+func (i *installer) Install(ctx context.Context) error {
+	err := i.checkReleasedVersions(ctx)
+	if err != nil {
+		return fmt.Errorf("check released versions: %w", err)
+	}
+	prj, err := i.loadRemoteProjectDefinition(ctx, i.version)
+	if err != nil {
+		return fmt.Errorf("remote: %w", err)
+	}
+	err = prj.EnsureFoldersExist(ctx)
+	if err != nil {
+		return fmt.Errorf("folders: %w", err)
+	}
+	prj.folder = PathInLabs(ctx, i.name)
+	err = i.authenticate(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("auth: %w", err)
+	}
+	err = i.downloadLibrary(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("lib: %w", err)
+	}
+	err = i.setupPythonVirtualEnvironment(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("python: %w", err)
+	}
+	err = i.recordVersion(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("record version: %w", err)
+	}
+	err = i.runInstallHook(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("installer: %w", err)
+	}
+	return nil
+}
+
+func (i *installer) Upgrade(ctx context.Context) error {
+	labsYml := PathInLabs(ctx, i.name, "lib", "labs.yml")
+	prj, err := Load(labsYml)
+	if err != nil {
+		return fmt.Errorf("load: %w", err)
+	}
+	err = i.checkReleasedVersions(ctx)
+	if err != nil {
+		return fmt.Errorf("check latest: %w", err)
+	}
+	err = i.downloadLibrary(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("lib: %w", err)
+	}
+	err = i.recordVersion(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("record version: %w", err)
+	}
+	err = i.runInstallHook(ctx, prj)
+	if err != nil {
+		return fmt.Errorf("installer: %w", err)
+	}
+	return nil
+}
+
+func (i *installer) warningf(text string, v ...any) {
+	// TODO: make bold
+	i.cmd.PrintErrf(text, v...)
+}
+
+func (i *installer) recordVersion(ctx context.Context, prj *Project) error {
+	return prj.writeVersionFile(ctx, i.version)
+}
+
+func (i *installer) checkReleasedVersions(ctx context.Context) error {
+	cacheDir := PathInLabs(ctx, i.name, "cache")
+	err := os.MkdirAll(cacheDir, 0o600)
+	if err != nil {
+		return fmt.Errorf("make cache dir: %w", err)
+	}
+	versions, err := releases.NewReleaseCache("databrickslabs", i.name, cacheDir).Load(ctx)
+	if err != nil {
+		return fmt.Errorf("versions: %w", err)
+	}
+	for _, v := range versions {
+		if v.Version == i.version {
+			return nil
+		}
+	}
+	if i.version == "latest" && len(versions) > 0 {
+		log.Debugf(ctx, "Latest %s version is: %s", i.name, versions[0].Version)
+		i.version = versions[0].Version
+		return nil
+	}
+	i.warningf("[WARNING] Installing unreleased version: %s", i.version)
+	return nil
+}
+
+func (i *installer) authenticate(ctx context.Context, prj *Project) error {
+	var cfg authConfig
+	profileOverride := prj.profileOverride(i.cmd)
```

Comment:
Do we need cmd only to get the profile flag, or also for other stuff?

Might be a way to get this information without the cmd.